controllers/semeru_compiler.go

```diff
@@ -114,9 +120,104 @@ func (r *ReconcileWebSphereLiberty) reconcileSemeruCompiler(wlva *wlv1.WebSphere
+	if wlva.Status.References != nil {
+		if completedInstances, ok := wlva.Status.References[StatusReferenceSemeruInstancesCompleted]; ok {
+			if completedInstances != currentGeneration && !strings.Contains(completedInstances, currentGeneration) {
+				wlva.Status.References[StatusReferenceSemeruInstancesCompleted] += "," + currentGeneration
+				// Delete old Semeru Cloud Compiler instances
+				r.deleteCompletedSemeruInstances(wlva)
+			}
+		} else {
+			wlva.Status.References[StatusReferenceSemeruInstancesCompleted] = currentGeneration
+		}
+	}
 	return nil, ""
 }
 
+// Returns the one-based index generation indicated by .status.references.semeruGeneration if it exists, otherwise defaults to 1
+func getGeneration(wlva *wlv1.WebSphereLibertyApplication) string {
+	if wlva.Status.References != nil && wlva.GetSemeruCloudCompiler() != nil {
+		if semeruGeneration, ok := wlva.Status.References[StatusReferenceSemeruGeneration]; ok {
+			return semeruGeneration
+		}
+		wlva.Status.References[StatusReferenceSemeruGeneration] = fmt.Sprint(1)
+	}
+	return "1"
+}
+
+// Increments the generation number at .status.references.semeruGeneration if it exists, otherwise if possible, initializes the generation to 1
+func createNewSemeruGeneration(wlva *wlv1.WebSphereLibertyApplication) {
+	if wlva.Status.References != nil {
+		if semeruGeneration, ok := wlva.Status.References[StatusReferenceSemeruGeneration]; ok {
+			if generation, err := strconv.Atoi(semeruGeneration); err == nil {
+				wlva.Status.References[StatusReferenceSemeruGeneration] = fmt.Sprint(generation + 1)
+			} else {
+				wlva.Status.References[StatusReferenceSemeruGeneration] = fmt.Sprint(1)
+			}
+		} else {
+			wlva.Status.References[StatusReferenceSemeruGeneration] = fmt.Sprint(1)
+		}
+	}
+}
+
+// Deletes old Semeru Cloud Compiler instances that have been marked as completed (instances that underwent at least one reconcile)
+func (r *ReconcileWebSphereLiberty) deleteCompletedSemeruInstances(wlva *wlv1.WebSphereLibertyApplication) {
+	if semeruInstancesCompleted, ok := wlva.Status.References[StatusReferenceSemeruInstancesCompleted]; ok {
+		generationsMarkedForDeletion := make([]string, 0)
+		cmPresent, _ := r.IsGroupVersionSupported(certmanagerv1.SchemeGroupVersion.String(), "Certificate")
+		useCertManager := !r.IsOpenShift() || cmPresent
+
+		// For each completed Semeru Cloud Compiler generation
+		for _, completedGenerationStr := range strings.Split(semeruInstancesCompleted, ",") {
+			completedGeneration, _ := strconv.Atoi(completedGenerationStr)
+			currentGeneration, _ := strconv.Atoi(wlva.Status.References[StatusReferenceSemeruGeneration])
+
+			// Delete the older generation's resources and mark the status reference field for deletion
+			if completedGeneration < currentGeneration {
+				semeruLabels := map[string]string{
+					SemeruGenerationLabelName: completedGenerationStr,
```

Comment:
This should also include the label `"app.kubernetes.io/name"` label to avoid deleting the JIT server instances from other WebSphereLibertyApplications in the same namespace

Reply:
Done